apiserver/facades/client/application/application.go

```diff
@@ -1053,9 +1058,92 @@ func (api *APIBase) applicationSetCharm(
 		StorageConstraints: stateStorageConstraints,
 		EndpointBindings:   params.EndpointBindings,
 	}
+
+	// Disallow downgrading from a v2 charm to a v1 charm.
+	oldCharm, _, err := params.Application.Charm()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if charm.MetaFormat(oldCharm) >= charm.FormatV2 && charm.MetaFormat(newCharm) == charm.FormatV1 {
+		return errors.New("cannot downgrade from v2 charm format to v1")
+	}
+
+	// If upgrading from a pod-spec (v1) charm to sidecar (v2), override the
+	// application's series to what it would be for a fresh sidecar deploy.
+	oldSeries := params.Application.Series()
+	if oldSeries == series.Kubernetes.String() && charm.MetaFormat(newCharm) >= charm.FormatV2 && corecharm.IsKubernetes(newCharm) {
+		// Disallow upgrading from a v1 DaemonSet or Deployment type charm
+		// (only StatefulSet is supported in v2 right now).
+		deployment := oldCharm.Meta().Deployment
+		if deployment != nil && deployment.DeploymentType != charm.DeploymentStateful {
+			return errors.Errorf("cannot upgrade from v1 %s deployment to v2", deployment.DeploymentType)
+		}
+
+		modelConfig, err := api.model.ModelConfig()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var supported []string
+		for _, base := range newCharm.Manifest().Bases {
```

Comment:
This can return nil in some situations. We should guard against panics here.

Reply:
But this is inside an if statement which checks `charm.MetaFormat(newCharm) >= charm.FormatV2` which guarantees Manifest() is not nil right?